price-feeder/oracle/provider/osmosisv2.go

```diff
@@ -0,0 +1,415 @@
+		osmosisV2Pair := currencyPairToOsmosisV2Pair(pair)
+		if msg, ok := messageResp[osmosisV2Pair]; ok {
+			switch v := msg.(type) {
+			// ticker response
+			case map[string]interface{}:
+				tickerString, _ := json.Marshal(v)
+				tickerErr = json.Unmarshal(tickerString, &tickerResp)
+				if tickerErr != nil {
+					p.logger.Error().
+						Int("length", len(bz)).
+						AnErr("ticker", tickerErr).
+						Msg("Error on receive message")
+
+					return
+				}
+				p.setTickerPair(
+					osmosisV2Pair,
+					tickerResp,
+				)
+				telemetryWebsocketMessage(ProviderOsmosisV2, MessageTypeTicker)
+
+				return
+
+			// candle response
+			case []interface{}:
+				// use latest candlestick in list
+				candleString, _ := json.Marshal(v[len(v)-1].(map[string]interface{}))
+				candleErr = json.Unmarshal(candleString, &candleResp)
+				if candleErr != nil {
+					p.logger.Error().
+						Int("length", len(bz)).
+						AnErr("candle", candleErr).
+						Msg("Error on receive message")
+
+					return
+				}
+				p.setCandlePair(
+					osmosisV2Pair,
+					candleResp,
+				)
+				telemetryWebsocketMessage(ProviderOsmosisV2, MessageTypeCandle)
+
+				return
+			}
+		}
+	}
+
+	p.logger.Error().
+		Int("length", len(bz)).
+		AnErr("message", messageErr).
+		Msg("Error on receive message")
+}
+
+// ping to check websocket connection
+func (p *OsmosisV2Provider) ping() error {
+	return p.wsClient.WriteMessage(websocket.PingMessage, ping)
+}
+
+func (p *OsmosisV2Provider) setTickerPair(symbol string, tickerPair OsmosisV2Ticker) {
+	p.mtx.Lock()
+	defer p.mtx.Unlock()
+
+	price, err := sdk.NewDecFromStr(tickerPair.Price)
+	if err != nil {
+		p.logger.Warn().Err(err).Msg("osmosisv2: failed to parse ticker price")
+		return
+	}
+	volume, err := sdk.NewDecFromStr(tickerPair.Volume)
+	if err != nil {
+		p.logger.Warn().Err(err).Msg("osmosisv2: failed to parse ticker volume")
+		return
+	}
+
+	p.tickers[symbol] = types.TickerPrice{
+		Price:  price,
+		Volume: volume,
+	}
+}
+
+func (p *OsmosisV2Provider) setCandlePair(symbol string, candlePair OsmosisV2Candle) {
+	p.mtx.Lock()
+	defer p.mtx.Unlock()
+
+	close, err := sdk.NewDecFromStr(candlePair.Close)
+	if err != nil {
+		p.logger.Warn().Err(err).Msg("osmosisv2: failed to parse candle close")
+		return
+	}
+	volume, err := sdk.NewDecFromStr(candlePair.Volume)
+	if err != nil {
+		p.logger.Warn().Err(err).Msg("osmosisv2: failed to parse candle volume")
+		return
+	}
+	candle := types.CandlePrice{
+		Price:  close,
+		Volume: volume,
+		// convert seconds -> milli
+		TimeStamp: SecondsToMilli(candlePair.EndTime),
+	}
+
+	staleTime := PastUnixTime(providerCandlePeriod)
+	candleList := []types.CandlePrice{}
+	candleList = append(candleList, candle)
+
+	for _, c := range p.candles[symbol] {
+		if staleTime < c.TimeStamp {
+			candleList = append(candleList, c)
+		}
+	}
+
+	p.candles[symbol] = candleList
+}
+
+func (p *OsmosisV2Provider) handleWebSocketMsgs(ctx context.Context, pairs ...types.CurrencyPair) {
```

Comment:
You could use the WebsocketController instead of duplicating this code. (https://github.com/umee-network/umee/blob/main/price-feeder/oracle/provider/mexc.go#L114). There is already an issue to convert the rest of the providers over so we can also wait until then.

Reply:
Let's definitely use the controller as we add new providers, less cleanup for later

Reply:
Done 👍